controllers/clustertopology_controller_compute.go

```diff
@@ -140,9 +142,132 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
+	infra, err := r.getReference(ctx, cluster.Spec.InfrastructureRef)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read %s", cluster.Spec.InfrastructureRef.Kind)
+	}
+	return infra, nil
+}
+
+// getCurrentControlPlaneState returns information on the ControlPlane being used by the Cluster; If a references is not found,
+// an error is thrown. If the ControlPlane requires MachineInfrastructure according to its ClusterClass an error will be
+// thrown if the ControlPlane has no MachineTemplates.
+func (r *ClusterTopologyReconciler) getCurrentControlPlaneState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (controlPlaneTopologyState, error) {
+	cp, err := r.getReference(ctx, cluster.Spec.ControlPlaneRef)
+	if err != nil {
+		return controlPlaneTopologyState{}, errors.Wrapf(err, "failed to read %s %s", cluster.Spec.ControlPlaneRef.Kind, cluster.Spec.ControlPlaneRef.Name)
+	}
+	log.Print(class.Spec.ControlPlane.MachineInfrastructure)
+	// Some ControlPlane providers may not require MachineInfrastructure to run. This check returns early if the field
+	// indicating MachineInfrastructure is required is not found in the clusterClass of the given cluster.
+	if class.Spec.ControlPlane.MachineInfrastructure == nil {
+		return controlPlaneTopologyState{
+			cp,
+			nil,
+		}, nil
+	}
+	cpInfraRef, err := getNestedRef(cp, "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", "ControlPlane", cluster.GetName())
+	}
+	infra, err := r.getReference(ctx, cpInfraRef)
+	if err != nil {
+		return controlPlaneTopologyState{cp, nil}, errors.Wrapf(err, "failed to get InfrastructureMachineTemplate reference for %s, %s", "ControlPlane", cluster.GetName())
+	}
+	return controlPlaneTopologyState{
+		cp,
+		infra,
+	}, nil
+}
+
+// getCurrentMachineDeploymentState queries for all MachineDeployments and filters them for their linked Cluster and
+// whether they are managed by a ClusterClass using labels. A cluster may have zero or more MachineDeployments. Zero is
+// expected on first reconcile. If MachineDeployments are found for the cluster Infrastructure and Bootstrap references
+// are inspected. Where these are not found the function will throw an error.
+func (r *ClusterTopologyReconciler) getCurrentMachineDeploymentState(ctx context.Context, cluster *clusterv1.Cluster) ([]machineDeploymentTopologyState, error) {
+	mDeploymentState := make([]machineDeploymentTopologyState, 0)
+
+	md := &clusterv1.MachineDeploymentList{}
+	err := r.Client.List(ctx, md, client.MatchingLabels{
+		clusterv1.ClusterLabelName:         cluster.Name,
+		clusterv1.ClusterTopologyLabelName: "",
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read MachineDeployments for managed topology")
+	}
+	for _, m := range md.Items {
```

Comment:
```suggestion

	for _, m := range md.Items {
```
I personally would also add a new line here